common/src/main/java/com/wynntils/core/chat/ChatModel.java

```diff
@@ -110,19 +109,19 @@ public static void onChatReceived(ChatPacketReceivedEvent e) {
     }
 
     private static void handleMultilineMessage(Component message) {
-        String msg = ComponentUtils.getCoded(message);
-
-        List<String> lines = new LinkedList<>(Arrays.asList(msg.split("\\n")));
+        List<Component> lines = new LinkedList<>(message.getSiblings());
```

Comment:
I'm seriously not really understanding how this can work. From what I remember, the "\n" was embedded in the middle of messages, and the "sibling" relationship segmented the message according to formatting instead. That was kind of the reason I started using formatted strings instead of Components. 

How do you ensure that `lines` contains just "lines"?

Reply:
From what I've seen, when the chat contents are resent, it is a huge, empty TextComponent, with each message being a sibling (these components might have additional siblings, but we do not care about them). As we do not split on "\n",  I needed to change a few places where we expected empty lines, etc. This seemed to work fine.

Reply:
This might have been a change with 2.0 then. I think I'll need to verify this myself. It is indeed very welcome if this is the case. I am 99% certain it did not look like that when I created this functionality.

Reply:
Well, it is still the case that `\n` are embedded. But perhaps it does not matter:

```
TextComponent{text='
                       ', siblings=[TextComponent{text='Welcome to Wynncraft!
', siblings=[], style=Style{ color=dark_gray, bold=true, italic=null, underlined=null, strikethrough=null, obfuscated=null, clickEvent=null, hoverEvent=null, insertion=null, font=minecraft:default}}, TextComponent{text='               ', siblings=[], style=Style{ color=gray, bold=null, italic=null, underlined=null, strikethrough=null, obfuscated=null, clickEvent=null, hoverEvent=null, insertion=null, font=minecraft:default}}, TextComponent{text='play.wynncraft.com', siblings=[], style=Style{ color=gray, bold=null, italic=null, underlined=null, strikethrough=null, obfuscated=null, clickEvent=ClickEvent{action=OPEN_URL, value='http://play.wynncraft.com'}, hoverEvent=null, insertion=null, font=minecraft:default}}, TextComponent{text=' ', siblings=[], style=Style{ color=gray, bold=null, italic=null, underlined=null, strikethrough=null, obfuscated=null, clickEvent=null, hoverEvent=null, insertion=null, font=minecraft:default}}, TextComponent{text='-/-', siblings=[], style=Style{ color=dark_gray, bold=null, italic=null, underlined=null, strikethrough=null, obfuscated=null, clickEvent=null, hoverEvent=null, insertion=null, font=minecraft:default}}, TextComponent{text=' wynncraft.com

', siblings=[], style=Style{ color=gray, bold=null, italic=null, underlined=null, strikethrough=null, obfuscated=null, clickEvent=null, hoverEvent=null, insertion=null, font=minecraft:default}}, TextComponent{text='                       ', siblings=[], style=Style{ color=gray, bold=null, italic=null, underlined=null, strikethrough=null, obfuscated=null, clickEvent=null, hoverEvent=null, insertion=null, font=minecraft:default}}, TextComponent{text='Wynncraft 2.0 is OUT!
', siblings=[], style=Style{ color=gray, bold=true, italic=null, underlined=null, strikethrough=null, obfuscated=null, clickEvent=null, hoverEvent=null, insertion=null, font=minecraft:default}}, TextComponent{text='                     ', siblings=[], style=Style{ color=gray, bold=null, italic=null, underlined=null, strikethrough=null, obfuscated=null, clickEvent=null, hoverEvent=null, insertion=null, font=minecraft:default}}, TextComponent{text='Read about the new abilities
', siblings=[], style=Style{ color=gray, bold=null, italic=null, underlined=null, strikethrough=null, obfuscated=null, clickEvent=null, hoverEvent=null, insertion=null, font=minecraft:default}}, TextComponent{text='                         ', siblings=[], style=Style{ color=gray, bold=null, italic=null, underlined=null, strikethrough=null, obfuscated=null, clickEvent=null, hoverEvent=null, insertion=null, font=minecraft:default}}, TextComponent{text='and archetypes ', siblings=[], style=Style{ color=gray, bold=null, italic=null, underlined=null, strikethrough=null, obfuscated=null, clickEvent=null, hoverEvent=null, insertion=null, font=minecraft:default}}, TextComponent{text='here!', siblings=[], style=Style{ color=gray, bold=null, italic=null, underlined=true, strikethrough=null, obfuscated=null, clickEvent=null, hoverEvent=null, insertion=null, font=minecraft:default}}, TextComponent{text='

', siblings=[], style=Style{ color=gray, bold=null, italic=null, underlined=null, strikethrough=null, obfuscated=null, clickEvent=null, hoverEvent=null, insertion=null, font=minecraft:default}}], style=Style{ color=null, bold=null, italic=null, underlined=null, strikethrough=null, obfuscated=null, clickEvent=null, hoverEvent=null, insertion=null, font=minecraft:default}}
```

Reply:
When the siblings have newlines, then it is some special Wynncraft message. I think it is fine to keep them as a "bunch" (for our purpose of quest dialogue parsing).

Reply:
All "normal" chat history (user sent messages) seem to have one sibling per line. So these "multiline" advertising messages perhaps don't matter that they are several lines long.

But... I wonder what happens when they are starting to creep out of the history range. Can it lead us to fail to detect the "last message"?

I am seriously worried a bit that this entire change is making the chat manager less robust. If nothing else, we will now be *relying* on the assumption that there will be basically one sibling per line (or several-lines-long-messages). If this breaks for some reason, due to chat message types we have not yet spotted, the entire logic can just come crumbling down, and deliver very weird results to the user.

Reply:
To feel 100% confident, I'd rather see that we transformed the TextComponent tree, so each top-level sibling contained exactly one line. For the normal chat messages, this is already the case, so we need not do anything there.

Reply:
This change might make us more reliant on Wynn in a way yes. But, we used to rely on newlines being in specific places, in a specific format (always newline before "Press shift to continue", etc.). I think this state of ChatModel is almost as "breakable" as the old version.

Reply:
> To feel 100% confident, I'd rather see that we transformed the TextComponent tree, so each top-level sibling contained exactly one line. For the normal chat messages, this is already the case, so we need not do anything there.

I am not sure doing this would be any better, as there is a really limited amount of messages which still have new lines, and even if they exist, they do not break our parsing system.

Reply:
Some pseudo code on how I'm thinking:
```
TextComponent newComponent = new MutableComponent();
for sibling : TextComponent.getSiblings() {
  String text = sibling.getString();
  if (text.indexOf("\n") == text.length()) {
    // all is fine
   newComponent.add(sibling);
    continue
  }
  // else we need to either split this sibling up, or join it with the next, or possiblt both
 doTrickyStuff();
}
textComponent = newComponent;
```
The only really complicated thing is the `doTrickyStuff()`....

Reply:
Also, technically, if we do this, this should be able to recurse on all siblings of all siblings. 

Still, I am not sure why we are trying to "optimize" the Component representation of a message (it feels like we are doing that). We could do this, but, either way, both the optimized and the normal component has the full information of a "message" (even if the said message is multi-line).

Reply:
> I am not sure doing this would be any better, as there is a really limited amount of messages which still have new lines, and even if they exist, they do not break our parsing system.

Maybe. I'm not fully convinced. I'm basically worrying about two things:

1) That there might be "special" messages that break this behavior, but that we have not yet seen. This can only be answered by running for a long time and seeing that it does not break.

2) That there is some edge case with how the "last stored chat line" and the  system for how new lines are detected with these ads spanning multiple lines. I cannot say for certain that it will break. But I *can* say that when I wrote the code I made the assumption that we got lines split on line breaks, and now that is not true anymore. So the logic needs to be scrutinized. And I think we need to make some practical tests, where we try to fill the chat so let's say just haft the first welcome or advertising message is left in the chat history, and then we go talk to an NPC, and see what happens.

3) There might be some point 3 as well, but I can't put my finger on it yet...

Reply:
Let me phrase things differently: I am not overly confident in the stability of this code. In contrast with e.g. the Container Query which I have a lot of faith in, this was tricky to get right, I made a lot of false starts and had to fix things when developing the code that I did not expect when I wrote it. So in my eyes it is very brittle. And I worry that this change will make it even more brittle. When I instead would like to make it more robust.

Reply:
It is not about optimizing storage, it is about storing the history of the chat on a per-line basis, since minecraft buffers N lines (I think it is 32, or something like that), and that the check-backwards-for-actually-new-messages work on a per-line basis. The disconnect between actual lines as a basic type (which my original code had) and siblings as a basic type, which this PR has, makes me worried that things can go wrong in ways that is hard to predict and hard to test.

Reply:
To put it bluntly:
```
        List<String> lines = new LinkedList<>(Arrays.asList(msg.split("\\n")));
```
was true, but

```
        List<Component> lines = new LinkedList<>(message.getSiblings());
```

is a lie. This is not a list of lines. It is a list of siblings. And that will have consequences. I like the idea to keep the messages as Components. But I'd *also* like to keep the idea that `lines` are, in fact, lines.

Reply:
I agree that this might not be the most stable/bug-free code we have. However, parsing chat data will always be somewhat uncertain. However, I do think that this PR is not that big of a downgrade to our code in terms of safety/stability. 

Yes, I am a bit paranoid about merging this PR. But, I don't think we can do this much better. Basically, the logic is quite simple now. Is current like a NPC Dialog, according to the pattern, or is it not? The other code is just handling the problem of Wynn re-sending messages.

Reply:
Overall, I am not even sure how the component should look like when we "squash" it.

<img width="489" alt="image" src="https://user-images.githubusercontent.com/49001742/191846028-5fd70fc3-bb12-4d6e-a780-32a9f9c3697c.png">

What do we do with the spaces that make this centered? Do we count the first line as line 1, and the 2-3 as a line again? I think you can see what the problem is. We cannot really turn this message into 1 line, without making assumptions. And, one of our goal is to have less assumptions in the code. So we just fix a problem by creating essentially the same problem somewhere else.

Reply:
I've started trying to write a method for doing this reshuffling of the TextComponent.
Basically what I need to be happy is that this line is replaced with a 
`List<Component> lines = getLinesAsSiblings(message);`

I can help trying to write `getLinesAsSiblings` if you want.

Reply:
After writing the code, and having proper component - line relationship, do you want to put back the newline checks (is it even necessary if every quest message is its own sibling)?

Reply:
Writing the code is half the problem (obviously, it is tricky). I am just not sure this is any better (as I assume it won't be worse. Although, in my head, the more modifications we do to the strings, the more bugs we will introduce).

Reply:
I found another example of a message that do not follow the "one sibling per line rule":

```
[22:44:03] [Render thread/INFO] (Minecraft) [CHAT] §3[INFO§3]§b TheNelston rewarded §r§31024 Emeralds§r§b to TheNelston.
[22:44:03] [Render thread/INFO] (Minecraft) [CHAT] §r§3Rewards can be claimed in the Member Menu.
```

This was sent as a single sibling.

Reply:
You mentioned your concern was that, when writing the ChatModel code, you had the assumption that the `lines` list always contains 1 chat line.

What if, we remove this assumption, and re-factor the ChatModel around this (although, I am not sure if there is anything to rewrite), instead of messing with components?

Reply:
> I found another example of a message that do not follow the "one sibling per line rule":
> 
> ```
> [22:44:03] [Render thread/INFO] (Minecraft) [CHAT] §3[INFO§3]§b TheNelston rewarded §r§31024 Emeralds§r§b to TheNelston.
> [22:44:03] [Render thread/INFO] (Minecraft) [CHAT] §r§3Rewards can be claimed in the Member Menu.
> ```
> 
> This was sent as a single sibling.

Seems like relevant messages are "grouped" together. Maybe we could even use this whole sibling - parent relation to our benefit? Example: Let's say we want to filter out the ad. We check for the first line, and we can just remove it's siblings as well. Same with the guild rewards. No need for multiple regexes.

Reply:
Yes, but what happens the first line scroll of the screen, but the second is still there? And you enter a NPC dialogue...

Reply:
The chat system of Vanilla is still line-based. That's a hard fact that we need to relate to, and it determines how we must approach this, imho.

Reply:
> Yes, but what happens the first line scroll of the screen, but the second is still there? And you enter a NPC dialogue...

I think Wynn either sends the whole component again (and mc just forgets it) or neither is sent. I don't think we confirmed this is not the case?

Reply:
No, we have not tested this. Or at least, I have not.